nasa/ow_simulator
# Suggested ow_lander action refactoring

# Linked Issues:
For feedback on the approach to [OCEANWATER-904](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-904)

# Summary of Changes
* **ow_actions** python package created to facilitate OceanWATERS lander actions. This provides a framework for 
  * Creating action for OceanWATERS
  * Easily creating actions servers or action client scripts from said actions
  * Unifying the implementation of arm actions and encapsulating arm action code behind a simple interface
* **LighSetIntensity**, **CameraCapture**, **Stow**, and **Unstow** all converted to this framework to provide examples of its usage. 
* Reduced repeated code and logic processes throughout arm action server implementation.

# Temporary Caveats
* Deactivated all other arm actions in this branch. This is only to ensure there are no collisions between the new `arm_action_server_experimental.py` (temporary name) and `arm_action_server.py` (which has been deactivated).
* The stop feature for arm actions is not yet implemented into the new framework, so `stop_action_client.py` will do nothing. The feature will be implemented in a later merge onto the OW-904 branch.

# Code Review Guide
This PR includes some code copy-pasting from `ow_lander/scripts`, much of which is only different by an indent. This section tells you what you can ignore since GitHub's diff won't catch what is the same for these files. 

## action_trajectories.py -> trajectory_planner.py 
* The class was made into a singleton called `ArmTrajectoryPlanner`, and the method `cascade_plans`,  now `_cascade_plans`, was moved into the global scope.
* An `__init__` function was added to initialize the singleton state. This made many function parameters obsolete, so wherever `move_arm`, `move_grinder`, `robot`, or `moveit_fk` were passed into a function, those parameters have been removed and their references in the method replaced with their corresponding member variables. 
* The `sever_stop` parameter was also removed, but it's functionality has not yet been replaced (see Caveats)
* `compute_forward_kinematics` method added
* `plan_arm_to_target` method added
* `check_for_stop` appeared to do nothing, so it's logic was disabled. Now always returns False. Will be removed later.

## trajectory_async_execution.py -> trajectory_executor.py
* The class was made into a singleton called `ArmTrajectoryExecutor`.
* An `__init__` function was added to initialize the singleton state. All the logic in the `connect` function was moved into here. 
* `connect` member variable removed. If a connection fails in __init__ the top level script fails with an exception. 

## utils.py -> common.py 
* Added the Singleton metaclass.

## constants.py -> constants.py (moved into ow_actions)
No changes from the original!

# Testing
## Test Stow and Unstow
1. Launch simulation into any world. `roslaunch ow atacama_y1a.launch`
2. Open 4 other terminals that all source `devel/setup.bash`:
  - **1st terminal** enter `rostopic echo /Unstow/goal`
  - **2nd terminal** enter `rostopic echo /Unstow/feedback` 
  - **3rd terminal** enter `rostopic echo /Unstow/result`
  - **4th terminal**, call the unstow action with `rosrun ow_lander unstow_client.py`
3. You should see a single goal populate the **1st terminal**, and it'll look something like 
```
header: 
  seq: 1
  stamp: 
    secs: 1916525590
    nsecs: 720000029
  frame_id: ''
goal_id: 
  stamp: 
    secs: 1916525590
    nsecs: 720000029
  id: "/unstow_client_647975_1669927495337-1-1916525590.720"
goal: 
  goal: 0.0
---
```
You should see the **2nd terminal** publish a feedback at a rapid rate (100 Hz). It'll look like
```
header: 
  seq: 4200
  stamp: 
    secs: 1916525604
    nsecs: 755000114
  frame_id: ''
status: 
  goal_id: 
    stamp: 
      secs: 1916525590
      nsecs: 720000029
    id: "/unstow_client_647975_1669927495337-1-1916525590.720"
  status: 1
  text: "This goal has been accepted by the simple action server"
feedback: 
  current: 
    x: 0.7587987870086269
    y: 0.18979331807595998
    z: 0.8811694656849518
---
```
4. The arm should move into the unstowed position, and when it arrives you should see the following goal in the **3rd terminal**
```
header: 
  seq: 4
  stamp: 
    secs: 1916525605
    nsecs: 444999933
  frame_id: ''
status: 
  goal_id: 
    stamp: 
      secs: 1916525590
      nsecs: 720000029
    id: "/unstow_client_647975_1669927495337-1-1916525590.720"
  status: 3
  text: "Unstow arm trajectory succeeded"
result: 
  final: 
    x: 0.7355859052905875
    y: 0.2552613456452144
    z: 0.9050746780690109
---
```
5. In any terminal, call stow `rosrun ow_lander stow_client.py` and verify the arm returns to the stowed position. No need to check goal/feedback/result messages here. 
6. Repeat stows and unstows as many times as you'd like. 

## Test Camera Capture
The following can be done in the same simulation world. No need to check action goal/feedback/result topics. 
1. Move mast to look at ground `rosrun ow_lander antenna_pan_tilt_action_client.py 1.0 0.5`
2. Bring up the Image View container in RQT.
3. Capture a picture using whatever exposure you'd like `rosrun ow_lander camera_capture_client.py 0.1`
4. Verify an image appeared in RQT.

## Test Light Set Intensity
1. Move mast to project its lights on the lander body `rosrun ow_lander antenna_pan_tilt_action_client.py 0.0 1.0`
2. Change light intensity as you like `rosrun ow_lander light_set_intensity_client.py`
3. Verify the corresponding mast light change intensity levels in the Gazebo sim. 

## Test Client Help messages
Try the following commands and verify the resulting help messages makes sense. 
```rosrun ow_lander unstow_client.py --help```
```rosrun ow_lander stow_client.py --help```
```rosrun ow_lander camera_capture_client.py --help```
```rosrun ow_lander light_set_intensity_client.py --help```
